internal/manager/run.go

```diff
@@ -312,3 +319,61 @@ func readyzHandler(mgr manager.Manager, dataplaneSynchronizer IsReady) func(*htt
 		return nil
 	}
 }
+
+type healthCheckHandler struct {
+	lock         sync.RWMutex
+	healthzCheck healthz.Checker
+	readyzCheck  healthz.Checker
+}
+
+func (s *healthCheckHandler) getHealthzCheck() healthz.Checker {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	return s.healthzCheck
+}
+
+func (s *healthCheckHandler) getReadyzCheck() healthz.Checker {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	return s.readyzCheck
+}
+
+func (s *healthCheckHandler) setHealthzCheck(checker healthz.Checker) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.healthzCheck = checker
+}
+
+func (s *healthCheckHandler) setReadyzCheck(checker healthz.Checker) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.readyzCheck = checker
+}
+
+func (s *healthCheckHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	var check healthz.Checker
+	switch req.URL.Path {
+	case "/healthz":
+		check = s.getHealthzCheck()
+	case "/readyz":
+		check = s.getReadyzCheck()
+	}
+
+	if check == nil {
+		rw.WriteHeader(http.StatusNotFound)
+		_, _ = rw.Write([]byte("not found"))
+		return
+	}
+
+	err := check(req)
+	if err != nil {
+		rw.WriteHeader(http.StatusInternalServerError)
+		_, _ = rw.Write([]byte(err.Error()))
```

Comment:
Original handlers also print the error in the response?

Reply:
https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/healthz/healthz.go 
The original healthz server responses:
`ok` + code `200` for check OK
`internal server error: <err>` + code `500` for error happened in checks
Let `http.NotFoundHadler` to return `404` for no check fuctions found.